adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java

```diff
@@ -316,6 +322,14 @@ private void doUploadMessage(final String tenant, final String registrationAsser
                     uploadHandler.fail("message not accepted by remote");
                 }
             }
+
+            // if client closes connection right after publishing a message, it happens that the closeHandler will be called
+            // before the assertion is put into the map (due to async way for getting assertion, creating sender and so on).
```

Comment:
my understanding is that you want to prevent that we keep the registration in the map in the case where the client has closed the connection before the registration assertion has been returned by Device Registration.

Shouldn't we simply introduce a corresponding `MqttEndpoint.isConnected` check in method `getRegistrationAssertion` right before we put the assertion to the map? We could then simply rely on the close handler, which I would prefer ...